core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Wraps a {@code WatchService} and allows paths to be registered.
+ */
+final class PropertiesFileWatcherRegistry implements AutoCloseable {
+    private static final Logger logger = LoggerFactory.getLogger(PropertiesFileWatcherRegistry.class);
+
+    @Nullable
+    private CompletableFuture<Void> future;
+    private final Map<Path, PropertiesFileWatcherContext> ctxRegistry = new ConcurrentHashMap<>();
+    private final WatchService watchService;
+    private final ExecutorService executor =
+            Executors.newSingleThreadExecutor(new ThreadFactoryBuilder().setDaemon(true).build());
+
+    /**
+     * Create a registry using the default file system.
+     */
+    PropertiesFileWatcherRegistry() {
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
```

Comment:
If I understood the NIO.2 API correctly, you have to use the `FileSystem` retrieved from `Path`:

```java
final Path path = Paths.get(uri);
watchService = path.getFileSystem().newWatchService();
```

This means, we may have more than one `WatchService` implementations depending on how a user configured the JRE, i.e. this class needs some refactoring to have different `PropertiesFileWatchRegistry`s for different `FileSystem`s.

Reply:
I'm still thinking about use-cases for this & how I can do this efficiently 🤔. Just leaving a comment to let you know I'm not disregarding this comment:)